prov/efa/src/rdm/efa_rdm_ope.c

```diff
@@ -1528,6 +1531,16 @@ int efa_rdm_ope_post_remote_write(struct efa_rdm_ope *ope)
 		if (OFI_UNLIKELY(!pkt_entry))
 			return -FI_EAGAIN;
 
+		if (ope->fi_flags & FI_INJECT) {
+			assert(ope->iov_count == 1);
+			assert(ope->total_len <= ep->inject_size);
+			copied = ofi_copy_from_hmem_iov(pkt_entry->wiredata + sizeof(struct efa_rdm_rma_context_pkt),
+				ope->total_len, FI_HMEM_SYSTEM, 0, ope->iov, ope->iov_count, 0);
+			assert(copied == ope->total_len);
+			ope->desc[0] = fi_mr_desc(pkt_entry->mr);
+			ope->iov[0].iov_base = pkt_entry->wiredata + sizeof(struct efa_rdm_rma_context_pkt);
+		}
+
 		write_once_len = MIN(ope->iov[iov_idx].iov_len - iov_offset,
 				    ope->rma_iov[rma_iov_idx].len - rma_iov_offset);
 		write_once_len = MIN(write_once_len, max_write_once_len);
```

Comment:
You copy the data from user buffer to the packet entry but the `efa_rdm_pke_init_write_context` call a few lines below is still using the user buffer

Reply:
I hack this by:
```
			ope->desc[0] = fi_mr_desc(pkt_entry->mr);
			ope->iov[0].iov_base = pkt_entry->wiredata + sizeof(struct efa_rdm_rma_context_pkt);
```

and that works b/c:
```
		efa_rdm_pke_init_write_context(
			pkt_entry, ope,
			(char *) ope->iov[iov_idx].iov_base + iov_offset,
			write_once_len, ope->desc[iov_idx],
			ope->rma_iov[rma_iov_idx].addr + rma_iov_offset,
			ope->rma_iov[rma_iov_idx].key);
```
passes in `ope->iov[iov_idx].iov_base + iov_offset` and `ope->desc[iov_idx]`, and it will only ever go around this loop once.

Reply:
The txe has a separate `cq_entry` field that has the buffer and length data. That information is copied to txe when the txe is created. So changing the `ope->iov` doesn't affect the completion and this OK to do https://github.com/ofiwg/libfabric/blob/d43c492191a4f8a90b4531fd25deba536a87ecfd/prov/efa/src/rdm/efa_rdm_ope.c#L85-L86